[WIP] Rewrite/Extend all resources

iranl opened the pull request:
Supersedes all earlier PR's by me. Implements all requested changes in those PR's.
Also fixes some other open issues.

Unfortunately, mainly because everything is moved to traits, it is now impossible to separate the changes into different PR's for each resource.

Changes (non exhaustive list, applies to alle resources)
- All fields (including fields that are only used once) moved to traits in the Fields namespace. Re-use traits whenever possible (same exact element name and type in Twinfield API).
- Created Enums for every field with an exhaustive list
- Make a PhpTwinfield object of elements whenever possible (in the same way the office code field is turned into an Office object)
- Implement all DOM attributes from element fields (name, shortname, dimensiontype, locked, inherit etc,)
- Rewrote all ApiConnectors, Mappers and Documents to mimic Transactions
- Added resources (see below)

| Component                                                                                                                         | get()              | listAll()          | send()             | delete()           |  Mapper            |
| ----------------------------------------------------------------------------------------------------------------------------------| :----------------: | :----------------: | :----------------: | :----------------: | :----------------: |
| [Activities](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Activities)                                | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Articles](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Articles)                                    | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Asset Methods](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/AssetMethods)                           | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Browse Data](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Request/BrowseData)                               | :white_check_mark: |                    |                    |                    | :white_check_mark: |
| Cash and Bank Books                                                                                                               |                    | :white_check_mark: |                    |                    | :white_check_mark: |
| [Cost Centers](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/CostCenters)                             | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| Countries                                                                                                                         |                    | :white_check_mark: |                    |                    | :white_check_mark: |
| [Currencies](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Currencies)                                | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Customers](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Customers)                                  | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Dimension Groups](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/DimensionGroups)                     | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Dimension Types](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/DimensionTypes)                       | :white_check_mark: | :white_check_mark: | :white_check_mark: |                    | :white_check_mark: |
| [Electronic Bank Statements](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Transactions/BankStatements)       |                    |                    | :white_check_mark: |                    |                    |
| [Fixed Assets](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/FixedAssets)                             | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [General Ledger Accounts](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/BalanceSheets)                | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Matching](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Miscellaneous/Matching)                              |                    |                    | :white_check_mark: |                    | :white_check_mark: |
| [Offices](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Offices)                                      | :white_check_mark: | :white_check_mark: |                    |                    | :white_check_mark: |
| Paycodes                                                                                                                          |                    | :white_check_mark: |                    |                    | :white_check_mark: |
| [Projects](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Projects)                                    | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Rates](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Rates)                                          | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| [Sales Invoices](https://c3.twinfield.com/webservices/documentation/#/ApiReference/SalesInvoices)                                 | :white_check_mark: | :white_check_mark: | :white_check_mark: |                    | :white_check_mark: |
| [Sales Invoice Types](https://c3.twinfield.com/webservices/documentation/#/ApiReference/SalesInvoices)                            |                    | :white_check_mark: |                    |                    | :white_check_mark: |
| [Suppliers](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Suppliers)                                  | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
|                                                                                                                                   |                    |                    |                    |                    |                    |
| <b>Transactions</b>                                                                                                               |                    |                    |                    |                    |                    |
[Bank Transactions](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Transactions/BankTransactions)                | :white_check_mark: |                    | :white_check_mark: | :white_check_mark: | :white_check_mark: |
[Cash Transactions](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Transactions/CashTransactions)                | :white_check_mark: |                    | :white_check_mark: | :white_check_mark: | :white_check_mark: |
[Journal Transactions](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Transactions/JournalTransactions)          | :white_check_mark: |                    | :white_check_mark: | :white_check_mark: | :white_check_mark: |
[Purchase Transactions](https://c3.twinfield.com/webservices/documentation/#/ApiReference/PurchaseTransactions)                     | :white_check_mark: |                    | :white_check_mark: | :white_check_mark: | :white_check_mark: |
[Sale Transactions](https://c3.twinfield.com/webservices/documentation/#/ApiReference/SalesTransactions)                            | :white_check_mark: |                    | :white_check_mark: | :white_check_mark: | :white_check_mark: |
|                                                                                                                                   |                    |                    |                    |                    |                    |
| [Users](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/Users)                                          | :white_check_mark: | :white_check_mark: |                    |                    | :white_check_mark: |
| User Roles                                                                                                                        |                    | :white_check_mark: |                    |                    | :white_check_mark: |
| [VAT](https://c3.twinfield.com/webservices/documentation/#/ApiReference/Masters/VAT)                                              | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| VAT Groups                                                                                                                        |                    | :white_check_mark: |                    |                    | :white_check_mark: |
| VAT Group Countries                                                                                                               |                    | :white_check_mark: |                    |                    | :white_check_mark: |


Breaking Changes
- Basically breaks everything because instead of strings in a lot of cases PhpTwinfield Objects or Enums are now expected instead. 

To Do
- Document breaking changes

iranl committed:
Update CustomerIntegrationTest.php

iranl committed:
Add files via upload

iranl committed:
Update CustomerApiConnectorTest.php

iranl committed:
Add files via upload

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Add files via upload

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update composer.json

iranl committed:
Update composer.json

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Add files via upload

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update invoiceSendRequest.xml

iranl committed:
Update invoiceFinalGetResponse.xml

iranl committed:
Update invoiceConceptGetResponse.xml

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update invoiceSendRequest.xml

iranl committed:
Add files via upload

iranl committed:
Add files via upload

iranl committed:
Update OfficeTest.php

iranl committed:
Update CustomerIntegrationTest.php

iranl committed:
Create customerListResponse.json

iranl committed:
Delete customerListResponse.xml

iranl committed:
Update CustomerIntegrationTest.php

iranl committed:
Update CustomerIntegrationTest.php

iranl committed:
Update .travis.yml

iranl committed:
Add files via upload

iranl committed:
Update .travis.yml

iranl committed:
Merge branch 'extend-all' into extend-all-testing

iranl committed:
Merge pull request #2 from iranl/extend-all-testing

Extend all testing

iranl committed:
Update .travis.yml

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #3 from iranl/extend-all-testing

Extend all testing

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update InvoiceMapper.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #4 from iranl/extend-all-testing

Extend all testing

iranl committed:
Update readme.md

remkobrenters commented:
Promising PR which we could use in our current projects. Any expectations when is to be reviewed?

iranl commented:
@remkobrenters As this PR is currently WIP i'm not expecting (although of course welcoming) a review at this time. 

I will try to complete the last outstanding functionality to-do's in the coming days (Complete mapper for Users and Office and send() function for Users). 

At that point I consider the PR feature complete and a preliminary review could be started at that time.

But a review will probably be a lot easier when I've completed the Unit/integration tests and documentation.

This will still take some time as i'm writing examples/docs and tests for 24(!) resources including examples of all API methods, all getters and all setters.

Op top of that there are breaking changes for 6 existing resources (Articles, Customers, Invoices, Bank Transactions, Suppliers and VAT) where replacement methods should be documented.

remkobrenters commented:
Thanks for the fast reply. We will keep an eye on this and maybe we can help if the fork gets stuck at some point to wrap it up in a production-ready state.

iranl committed:
Upload

iranl committed:
Update readme.md

iranl committed:
Update RenewAuthorization.php

willemstuursma commented:
@remkobrenters I don't mind some help reviewing. Its turning into a pretty big change.

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update readme.md

iranl committed:
Upload

RVxLab commented:
@iranl Does the finder wsdl always return its value as strings to your knowledge?

If so, this is currently an issue when requesting invoices. The invoicenumber property is expected to be a nullable int as per the your InvoiceNumberField Trait `setInvoiceNumber` signature.

See https://github.com/iranl/twinfield/blob/extend-all/src/Fields/Invoice/InvoiceNumberField.php

iranl commented:
@RVxLab The finder returns its items as an ArrayOfString, so yes I do believe its values are returned as strings. 

Maybe I'm not yet familiar enough with nullable type casting but I don't believe casting/juggling a numeric string value (which is the only possible value that the finder can return for this field, per the API documentation) to (nullable) int leads to anything other than an int with the correct value. This behavior was confirmed in my testing so far.

Can you elaborate in which real world situation you believe this will cause an issue?

As to your second suggestion -> I added the following method to BaseService (will add this soon to the PR), came across this as the easiest way to switch from __call to __soapcall when working with WSDL:

```php
public function __call($func, $args)
{
    return $this->__soapCall($func, $args);
}
```

RVxLab commented:
@iranl With the current implementation of the API, calling `InvoiceApiConnector::listAll()` will fetch results from both the new Invoice module and the old one. The old Invoice's invoiceNumber property is by my observations always a numeric string, which can be cast without any problems assuming strict mode is not enabled.

However, with Invoices from the new module, the invoices *can* be non-numeric, a notice with the message `A non well formed numeric value encountered`. Which in itself would not be an issue, except when using with a framework like Laravel, which converts all notices and warnings to errors.

While this can be worked around by using `error_reporting(E_ALL ^ E_NOTICE)`, it's not ideal.

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update LinesField.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update BaseTransactionLine.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update TransactionsDocument.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update CloseAndStartValueFields.php

iranl committed:
Update CloseAndStartValueFields.php

iranl committed:
Update CloseAndStartValueFields.php

iranl committed:
Update CloseAndStartValueFields.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update TransactionMapper.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update TransactionLineInterface.php

iranl committed:
Upload

iranl committed:
Update TransactionsDocument.php

iranl committed:
Upload

iranl committed:
Merge pull request #5 from iranl/extend-all-testing

Extend all testing

iranl commented:
@RVxLab Was not aware that Twinfield start returning results from the new modules and was also not aware that the new one allows non-numeric invoice numbers (contradictio in terminis). I just followed the API documentation in this one which also states it should be an integer.

Considering the above I changed the InvoiceNumber to a string instead of an int.

With the latest commits I also rewrote Transactions (almost completely) and BrowseData/Matching/ElectronicBankStatement (a bit) to match the rest of the components.

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #6 from iranl/extend-all-testing

Extend all testing

iranl committed:
Merge branch 'master' into extend-all

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
UploaD

iranl committed:
Update project-response.xml

iranl committed:
Delete ArticleApiConnectorTest.php

iranl committed:
Merge pull request #7 from iranl/extend-all-testing

Extend all testing

iranl committed:
Upload

iranl committed:
Update SupplierDocumentUnitTest.php

iranl committed:
Update SupplierDocumentUnitTest.php

iranl committed:
Update SupplierDocumentUnitTest.php

iranl committed:
Update SupplierDocumentUnitTest.php

willemstuursma commented:
To be honest @iranl, the amount of changes is a bit overwhelming for me (653 files changed). 

Looking at all the comments, it looks like its all moving into the right direction. But it will be hard for me to check it all. 

What would be some good ways to proceed?

iranl commented:
@willemstuursma Agree that it got a little out of hand. You kind of called it upon yourself when you requested me to do it right (e.g. traits/enums/typehints) though ;).

How about we break it down  into seperate PR chuncks as follows:

NOTE: FIRST PART OF THIS LIST HAS NO IMPACT ON ANY EXISTING FUNCTIONALITY
- Add additions to current base files (e.g. BaseApiConnector, BaseDocuments, BaseMapper)

- Other small modifications that have no effect on existing functionality

- Fields (account for half of all changes files, but are in large parts just copies of existing functions from the individual objects and can be devided into groups that are almost identical)
 - - Add String field traits
- - Add Int field traits
- - Add Bool field traits
- - Add Float field traits
- - Add Date/DateTime field traits
- - Add Money field traits
- - Add Enum field traits
- - Add PHPTwinfield object field traits (such as office/customer fields)

- Add new Enums and modify existing Enums where necessary

- Add new resources one by one (whenever possible)

NOTE: BREAKING CHANGES STARTING HERE

- Modify existing resources one by one (whenever possible) + modifiying tests where needed

- Add examples

- Add documentation (e.g. readme, breaking)

- Remove all unnecessary functions and files that are left over

Should get it down to 10-20 file changes per PR, the exception being the traits, but those are known functions and mostly copies of each other

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update CustomerDocumentUnitTest.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Merge pull request #8 from iranl/extend-all-testing

Extend all testing

willemstuursma commented:
@iranl That seems like a sensible plan.

I would greatly appreciate if you can create the PRs for it.

Lets keep this PR open, so we can slowly see the diff getting smaller.

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update BaseMapper.php

iranl committed:
Update BaseMapper.php

iranl committed:
Update BaseMapper.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #10 from iranl/extend-all-testing

Extend all testing

iranl committed:
Update BaseMapper.php

iranl committed:
Update BaseMapper.php

iranl committed:
Base Modifications Change Requests 2

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update BaseObject.php

iranl committed:
Update BaseMapper.php

iranl committed:
Upload

iranl committed:
Update Util.php

iranl committed:
Merge pull request #11 from iranl/extend-all-testing

Extend all testing

iranl committed:
Update BaseMapper.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update BankTransactionIntegrationTest.php

iranl committed:
Update BankTransactionIntegrationTest.php

iranl committed:
Upload

iranl committed:
Update BankTransactionIntegrationTest.php

iranl committed:
Update BankTransactionIntegrationTest.php

iranl committed:
Update CustomerIntegrationTest.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update BankTransactionIntegrationTest.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #12 from iranl/extend-all-testing

Extend all testing

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update CustomerDocumentUnitTest.php

iranl committed:
Update CustomerDocumentUnitTest.php

iranl committed:
Upload

iranl committed:
Update CustomerDocumentUnitTest.php

iranl committed:
Update BaseMapper.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update SalesTransactionIntegrationTest.php

iranl committed:
Upload

iranl committed:
Update CustomersDocument.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update BaseIntegrationTest.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update InvoiceIntegrationTest.php

iranl committed:
Update OfficeIntegrationTest.php

iranl committed:
Update OfficeIntegrationTest.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #13 from iranl/extend-all-testing

Extend all testing

iranl committed:
Update readme.md

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #14 from iranl/extend-all-testing

Extend all testing

iranl committed:
Upload

iranl committed:
Merge pull request #15 from iranl/extend-all-testing

Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update RateApiConnectorTest.php

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Update HasEqualInterface.php

iranl committed:
Upload

iranl committed:
Update VatCodeApiConnectorTest.php

iranl committed:
Upload

iranl committed:
Update AssetMethodApiConnectorTest.php

iranl committed:
Merge pull request #16 from iranl/extend-all-testing

Extend all testing

iranl committed:
Update readme.md

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #17 from iranl/extend-all-testing

Extend all testing

iranl committed:
Upload

iranl committed:
Upload

iranl committed:
Merge pull request #18 from iranl/extend-all-testing

Extend all testing

iranl committed:
Merge branch 'master' into extend-all

iranl committed:
Upload

iranl committed:
Merge pull request #19 from iranl/extend-all-testing

Upload

iranl committed:
Update readme.md
